shell: Improve handling of log messages (#12125)
zephyrproject-rtos/zephyr: doc/subsystems/shell/shell.rst

	  params  ping

Shell as the logger backend
***************************

Shell instance can act as the :ref:`logger` backend. Shell ensures that log

> messages are correctly mulitplexed with shell output. Log messages from logger

Reviewer: ```suggestion
 messages are correctly multiplexed with shell output. Log messages from logger
```

> thread are enqueued and processed in the shell thread. Logger thread will block

Reviewer 2: What if logs are processed from user thread or by direct calls to LOG_PROCESS()?

Author: this context still can be called logger thread context. Do you consider case when LOG_PROCESS() is called from context that cannot be blocked? If so then I can add assert to that call to ensure that message is passed from thread context.

for configurable amount of time if queue is full, blocking logger thread context
for that time. Oldest log message is removed from the queue after timeout and

> new message is enqueued. Use the ``shell stats show`` command to retrieve

Reviewer 3: I'd like to add that if user needs to check lost messages on their own, that's not ok. The fact that messages are being dropped should be actively shown to the user (as in #12062).

> new message is enqueued. `shell stats show` command can be used to retrieve

Reviewer: ```suggestion
 new message is enqueued. Use the ``shell stats show`` command to retrieve the
```

> number of log message dropped by the shell instance. Log queue size and timeout

Reviewer: ```suggestion
 number of log messages dropped by the shell instance. Log queue size and timeout
```

are :c:macro:`SHELL_DEFINE` arguments.

.. warning::
	Enqueueing timeout must be carefully set if multiple logger backends are

> 	used in the system. If timeout is set high to the shell instance which

Reviewer: If I understand this right, how about:
```
.. warning::
   Enqueuing timeout must be set carefully when multiple
   backends are used in the system. The shell instance could
   have a slow transport or could block, for example, by a
   UART with hardware flow control. If timeout is set too high,
   the logger thread could be blocked and impact other logger
   backends.
```
